src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -0,0 +1,200 @@
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.ServerPort))
+                {
+                    AddToList(w3cList, nameof(ConnectionInfo.LocalPort), connectionInfo.LocalPort.ToString(CultureInfo.InvariantCulture));
+                }
+            }
+
+            if ((W3CLoggingFields.Request & options.LoggingFields) != W3CLoggingFields.None)
+            {
+                var request = context.Request;
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.ProtocolVersion))
+                {
+                    AddToList(w3cList, nameof(request.Protocol), request.Protocol);
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.Method))
+                {
+                    AddToList(w3cList, nameof(request.Method), request.Method);
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.UriStem))
+                {
+                    AddToList(w3cList, nameof(request.Path), request.Path.Value);
+                }
+
+                if (options.LoggingFields.HasFlag(W3CLoggingFields.UriQuery))
+                {
+                    AddToList(w3cList, nameof(request.QueryString), request.QueryString.Value);
+                }
+
+                if ((W3CLoggingFields.RequestHeaders & options.LoggingFields) != W3CLoggingFields.None)
+                {
+                    var headers = request.Headers;
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.Host))
+                    {
+                        if (headers.TryGetValue(HeaderNames.Host, out var host))
+                        {
+                            AddToList(w3cList, HeaderNames.Host, host.ToString());
+                        }
+                    }
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.Referer))
+                    {
+                        if (headers.TryGetValue(HeaderNames.Referer, out var referer))
+                        {
+                            AddToList(w3cList, HeaderNames.Referer, referer.ToString());
+                        }
+                    }
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.UserAgent))
+                    {
+                        if (headers.TryGetValue(HeaderNames.UserAgent, out var agent))
+                        {
+                            AddToList(w3cList, HeaderNames.UserAgent, agent.ToString());
+                        }
+                    }
+
+                    if (options.LoggingFields.HasFlag(W3CLoggingFields.Cookie))
+                    {
+                        if (headers.TryGetValue(HeaderNames.Cookie, out var cookie))
+                        {
+                            AddToList(w3cList, HeaderNames.Cookie, cookie.ToString());
+                        }
+                    }
+                }
+            }
+
+            var response = context.Response;
+
+            await _next(context);
```

Comment:
Should we log something when next throws? I'm thinking we should probably log a 500 and rethrow, right?

Reply:
If _next throws, catch it, log whatever I have already, rethrow